internal/controller/bucket/update.go

```diff
@@ -0,0 +1,175 @@
+package bucket
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"
+
+	"github.com/pkg/errors"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/linode/provider-ceph/apis/provider-ceph/v1alpha1"
+	apisv1alpha1 "github.com/linode/provider-ceph/apis/v1alpha1"
+	"github.com/linode/provider-ceph/internal/backendstore"
+	s3internal "github.com/linode/provider-ceph/internal/s3"
+	"github.com/linode/provider-ceph/pkg/utils"
+)
+
+//nolint:gocyclo,cyclop // Function requires numerous checks.
+func (c *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	bucket, ok := mg.(*v1alpha1.Bucket)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotBucket)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.operationTimeout)
+	defer cancel()
+
+	if utils.IsHealthCheckBucket(bucket) {
+		c.log.Info("Update is NOOP for health check bucket - updates performed by heath-check-controller", "bucket", bucket.Name)
+
+		return managed.ExternalUpdate{}, nil
+	}
+
+	if bucket.Spec.Disabled {
+		c.log.Info("Bucket is disabled - remove any existing buckets from backends", "bucket name", bucket.Name)
+
+		return managed.ExternalUpdate{}, c.Delete(ctx, mg)
+	}
+
+	if len(bucket.Spec.Providers) == 0 {
+		bucket.Spec.Providers = c.backendStore.GetAllActiveBackendNames()
+	}
+
+	if err := c.updateAll(ctx, bucket); err != nil {
+		return managed.ExternalUpdate{}, err
+	}
+
+	err := c.updateObject(ctx, bucket, func(origBucket, bucket *v1alpha1.Bucket) bool {
+		bucket.Status.Conditions = origBucket.Status.Conditions
+		bucket.Status.AtProvider.BackendStatuses = origBucket.Status.AtProvider.BackendStatuses
+
+		return true
+	}, func(origBucket, bucket *v1alpha1.Bucket) bool {
+		bucket.Spec.Providers = origBucket.Spec.Providers
+
+		resourceUpToDate := true
```

Comment:
maybe rename this to something like `allBucketsReady`?... to avoid confusion with `updateObject` return values